torch/quantization/ns/graph_matcher.py

```diff
@@ -0,0 +1,315 @@
+    for part in fqn_parts:
+        cur_val = getattr(cur_val, part)
+    return cur_val
+
+class _NSGraphMatchableNodesIterator:
+    """
+    Iterates through the graph of gm, starting with the output nodes
+    and continuing backwards.
+    1. Returns matchable nodes, in order
+    2. Skips over non-matchable nodes
+    """
+    def __init__(
+        self,
+        gm: GraphModule,
+        non_matchable_functions: Set[Callable],
+        non_matchable_modules: Set[Callable],
+    ):
+        self.gm: GraphModule = gm
+        self.non_matchable_functions: Set[Callable] = non_matchable_functions
+        self.non_matchable_modules: Set[Callable] = non_matchable_modules
+        self.seen_nodes: Set[Node] = set()
+        self.stack: List[Node] = []
+        for start_node in _get_output_nodes(self.gm.graph):
+            self.stack.append(start_node)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self) -> Node:
+        """
+        Returns the next matchable node.
+        """
+        while len(self.stack) > 0:
+            cur_node = self.stack.pop()
+            if cur_node in self.seen_nodes:
+                continue
+            self.seen_nodes.add(cur_node)
+            # add args of previous nodes to stack
+            # TODO(future PR): handle kwargs as needed
+            for arg in cur_node.args:
+                if isinstance(arg, Node):
+                    self.stack.append(arg)
+                # TODO(future PR): handle other arg types such as Tuple, etc
+
+            # skip observers, etc
+            if not self._is_matchable(cur_node):
+                continue
+
+            return cur_node
+
+        raise StopIteration
+
+    def _is_matchable(self, node: Node) -> bool:
+        if node.op == 'call_function':
+            return not (node.target in self.non_matchable_functions)
+        elif node.op == 'call_module':
+            assert isinstance(node.target, str)
+            # target_mod = getattr(self.gm, node.target)
+            target_mod = _getattr_from_fqn(self.gm, node.target)
+            return not \
+                any(isinstance(target_mod, t)  # type: ignore
+                    for t in self.non_matchable_modules)
+        else:
+            return False
+
+class GraphMatchingException(Exception):
+    """
+    Exception raised when two graphs cannot be matched.
+    """
+    pass
+
+def _node_a_related_to_b(
+    node_a: Node,
+    node_b: Node,
+    gm_a: GraphModule,
+    gm_b: GraphModule,
+    type_a_related_to_b: Set[Tuple[Callable, Callable]],
+) -> bool:
+    if node_a.op != node_a.op:
+        # for now, comparing call_module to call_function is not supported
+        # this can be added later if needed
+        return False
+
+    if node_a.op == 'call_function':
+        if node_a.target == node_b.target:
+            # nodes with equivalent targets always match (i.e. F.linear and F.linear)
+            return True
+        key = (node_a.target, node_b.target)
```

Comment:
Alternately if both node_a.target and node_b.target belong to the same set, it could work.

Reply:
sure, I don't feel strongly about how we implement this as long as it works